ring/replication_set.go

```diff
@@ -388,6 +389,111 @@ func DoUntilQuorumWithoutSuccessfulContextCancellation[T any](ctx context.Contex
 	return results, nil
 }
 
+// DoMultiUntilQuorumWithoutSuccessfulContextCancellation behaves similar to DoUntilQuorumWithoutSuccessfulContextCancellation
+// with the following exceptions:
+//
+//   - This function calls DoUntilQuorumWithoutSuccessfulContextCancellation for each input ReplicationSet and requires
+//     DoUntilQuorumWithoutSuccessfulContextCancellation to successfully run for each of them. Execution breaks on the
+//     first error returned by DoUntilQuorumWithoutSuccessfulContextCancellation on any ReplicationSet.
+//
+//   - This function requires that the callback function f always call context.CancelCauseFunc once done. Failing to
+//     cancel the context will leak resources.
+func DoMultiUntilQuorumWithoutSuccessfulContextCancellation[T any](ctx context.Context, sets []ReplicationSet, cfg DoUntilQuorumConfig, f func(context.Context, *InstanceDesc, context.CancelCauseFunc) (T, error), cleanupFunc func(T)) ([]T, error) {
+	if len(sets) == 0 {
+		return nil, errors.New("no replication sets")
+	}
+	if len(sets) == 1 {
+		return DoUntilQuorumWithoutSuccessfulContextCancellation[T](ctx, sets[0], cfg, f, cleanupFunc)
+	}
+
+	results, _, err := doMultiUntilQuorumWithoutSuccessfulContextCancellation[T](ctx, sets, cfg, f, cleanupFunc)
+	return results, err
+}
+
+// See DoMultiUntilQuorumWithoutSuccessfulContextCancellation().
+//
+// The returned context.Context is the internal context used by workers and it's used for testing purposes.
+func doMultiUntilQuorumWithoutSuccessfulContextCancellation[T any](ctx context.Context, sets []ReplicationSet, cfg DoUntilQuorumConfig, f func(context.Context, *InstanceDesc, context.CancelCauseFunc) (T, error), cleanupFunc func(T)) ([]T, context.Context, error) {
+	var (
+		returnResultsMx = sync.Mutex{}
+		returnResults   = make([]T, 0, len(sets)*len(sets[0].Instances)) // Assume all replication sets have the same number of instances.
+
+		returnErrOnce sync.Once
+		returnErr     error // The first error occurred.
+
+		workersGroup                 = sync.WaitGroup{}
+		workersCtx, cancelWorkersCtx = context.WithCancelCause(ctx)
+
+		inflightTracker = newInflightInstanceTracker(sets)
+	)
+
+	cancelWorkersCtxIfSafe := func() {
+		if inflightTracker.allInstancesCompleted() {
+			cancelWorkersCtx(errors.New("all requests completed"))
```

Comment:
perhaps worth noting that this error should never be seen ever anywhere. I wouldn't open a follow-up PR just for this but if you do end up with one you can add a comment or change the error string to make this more clear

Reply:
Agree, but it's handy to assert in tests the context was cancelled for the expected reason ([example](https://github.com/grafana/dskit/blob/main/ring/replication_set_test.go#L1575)).